File: static/js/fcm.js
let VAPID;

//Firebase 메시징 설정/초기 키 로드
function load_key() {
    if (typeof firebase === 'undefined') {
        console.error('Firebase SDK is not loaded. Make sure it\'s included before fcm.js');
        return;
    }

    fetch('/fcm/firebase_config')
        .then(response => response.json())
        .then(firebaseConfig => {
            VAPID = firebaseConfig.VAPID;
            if (!firebase.apps.length) {
                firebase.initializeApp(firebaseConfig);
            } else {
                firebase.app(); // 이미 초기화되어 있다면 기존 앱을 사용
            }

            if (!firebase.messaging.isSupported()) {
                console.error('This browser doesn\'t support Firebase Messaging.');
                return;
            }

            window.messagingInstance = firebase.messaging();
            initializeFCM();
        })
        .catch(error => console.error('Error loading Firebase config:', error));
}

// 서비스 워커를 등록 및 메시지 수신 시 알림을 표시
function initializeFCM() {
    if ('serviceWorker' in navigator) {
        navigator.serviceWorker.register('/firebase-messaging-sw.js')
            .then(function (registration) {
                console.log('Service Worker registered with scope:', registration.scope);

                //메시지 수신 후 알림 표시 (포그라운드)
                window.messagingInstance.onMessage((payload) => {
                    console.log('Message received. ', payload);

                    const notificationOptions = {
                        body: payload.notification.body,
                        icon: payload.notification.icon,
                        data: {
                            url: payload.data.url,
                            title: payload.notification.title,

                        }
                    };

                    const notification = new Notification(payload.notification.title, notificationOptions);

                    // 알림 클릭 시 특정 URL로 이동
                    notification.onclick = function(event) {
                        window.open('/routine_notice?title=' + event.target.data.title + '&body=' + event.target.data.body, '_blank');
                    };
                });
            }).catch(function (err) {
                console.log('Service Worker registration failed: ', err);
            });
    }
}

//로그인 시 메시징 토큰을 요청 및 저장
window.onLogin = function () {
    if (!window.messagingInstance) {
        console.error('Firebase messaging is not initialized');
        return;
    }
    window.messagingInstance.getToken({ vapidKey: VAPID })
        .then((currentToken) => {
            if (currentToken) {
                sendTokenToServer(currentToken);
            } else {
                console.log('No registration token available. Request permission to generate one.');
            }
        }).catch((err) => {
            console.error('An error occurred while retrieving token. ', err);
        });
}

//FCM 토큰 저장
function sendTokenToServer(token) {
    fetch('/fcm/save_fcm_token', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({ token: token }),
    })
        .then(response => response.json())
        .then(data => console.log('Token saved:', data))
        .catch((error) => console.error('Error saving token:', error));
}

// 로그아웃 시 FCM 토큰을 삭제
window.onLogout = function () {
    if (!window.messagingInstance) {
        console.error('Firebase messaging is not initialized');
        return;
    }

    window.messagingInstance.deleteToken().then(() => {
        console.log('Token deleted.');
    }).catch((err) => {
        console.error('Unable to delete token. ', err);
    });
}

// 페이지 로드 시 Firebase 초기화
document.addEventListener('DOMContentLoaded', load_key);

File: main/fcm.py
from flask import Blueprint, request, jsonify, current_app
import flask_login
from model.model_platform import User, db
from firebase_admin import messaging
import requests
from config import Config

fcm = Blueprint('fcm', __name__)

# FCM 토큰 구독 처리
@fcm.route('/subscribe', methods=['POST'])
def subscribe():
    data = request.get_json()
    token = data.get('token')
    return jsonify({"success": True})

# 주어진 토큰으로 FCM 알림 전송
@fcm.route('/send_notification', methods=['POST'])
def send_notification():
    data = request.get_json()
    token = data.get('token')
    message = data.get('message')

    headers = {
        'Authorization': 'key=' + Config.apiKey,
        'Content-Type': 'application/json'
    }

    payload = {
        'to': token,
        'notification': {
            'title': 'Notification Title',
            'body': message
        },
        'data': {
            'url': '/routine_notice'
        }
    }

    response = requests.post('https://fcm.googleapis.com/fcm/send', headers=headers, json=payload)

    if response.status_code == 200:
        return jsonify({"success": True}), 200
    else:
        return jsonify({"failed": True}), 500

# Firebase 키 반환
@fcm.route('/firebase_config')
def firebase_config():
    config = {
        "apiKey": Config.apiKey,
        "authDomain": Config.authDomain,
        "projectId": Config.projectId,
        "storageBucket": Config.storageBucket,
        "messagingSenderId": Config.messagingSenderId,
        "appId": Config.appId,
        "VAPID":Config.VAPID,
        "measurementId": Config.measurementId,
    }
    return jsonify(config)

# FCM 토큰을 저장
@fcm.route('/save_fcm_token', methods=['POST'])
def save_fcm_token():
    data = request.json
    token = data.get('token')
    user_id = flask_login.current_user.UserID
    
    user = User.query.get(user_id)
    if user:
        user.FCMToken = token
        db.session.commit()
        return jsonify({"message": "Token saved successfully"}), 200
    return jsonify({"error": "User not found"}), 404
